### PR TITLE
feat: improve FinalityProofProvider api

### PR DESCRIPTION
**Pull request by yjhmelody**
It will be helpful if export the low level api for grandpa.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2382141

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2382156

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2382231

File: client/finality-grandpa/src/finality_proof.rs [hunk 1]
```diff
@@ -146,11 +147,14 @@ pub enum FinalityProofError {
 	Client(#[from] sp_blockchain::Error),
 }
 
-fn prove_finality<Block, B>(
+/// Prove finality for the given block number by returning a Justification for the last block of
+/// the authority set.
+pub fn prove_finality<Block, B>(
```
Comment (bkchr) [inline client/finality-grandpa/src/finality_proof.rs:152]: This should be kept private and you should add a function to `SharedAuthoritySet`.

Please also document what `collect_unknown_headers` does.

Comment (yjhmelody) [reply client/finality-grandpa/src/finality_proof.rs]: Ok, maybe user should impl this logic outside substrate by theirself.

Comment (yjhmelody) [reply client/finality-grandpa/src/finality_proof.rs]: Maybe I should close this PR.

**Review by andresilva — APPROVED** [review]
LGTM. If we want to make `prove_finality` public then we should probably not expose the end-user to having to deal with providing `AuthoritySetChanges`. We'd probably want to create another function that only takes as input the client/backend, the block to prove finality and the boolean to decide whether to collect unknown headers. Internally we'd need to read the `SharedAuthoritySet` from the database and extract the `AuthoritySetChanges` from there (https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/aux_schema.rs#L317).

**Comment by yjhmelody** [comment]
@andresilva Hi, I think adding one more method to `FinalityProofProvider` is enough. 
But I wonder why encode `justification` twice?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2387878

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2387882

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2387871

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2387872

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2387873

File: client/finality-grandpa/src/finality_proof.rs [hunk 2]
```diff
@@ -214,8 +243,8 @@ where
 		},
 	};
 
-	// Collect all headers from the requested block until the last block of the set
-	let unknown_headers = {
+	let unknown_headers = if collect_unknown_headers {
+		// Collect all headers from the requested block until the last block of the set
```
[the remark this answers is not in the document]
Comment (yjhmelody) [reply client/finality-grandpa/src/finality_proof.rs:240]: Updated

File: client/finality-grandpa/src/finality_proof.rs [hunk 3]
```diff
@@ -63,6 +63,15 @@ pub struct FinalityProofProvider<BE, Block: BlockT> {
...
 
+impl<BE, Block: BlockT> Clone for FinalityProofProvider<BE, Block> {
+	fn clone(&self) -> Self {
+		Self {
+			backend: self.backend.clone(),
+			shared_authority_set: self.shared_authority_set.clone(),
+		}
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (yjhmelody) [reply client/finality-grandpa/src/finality_proof.rs]: Updated

**Review by davxy — APPROVED** [review]

**Review by bkchr — APPROVED** [review]
Thank you!

**Comment by bkchr** [comment]
bot merge

**Pull request merged**